src/Featurizers/LagLeadOperatorFeaturizer.h

```diff
@@ -0,0 +1,224 @@
+    void execute_impl(typename BaseType::InputType const &input, typename BaseType::CallbackFunction const &callback) override {
+        execute_helper(nonstd::optional<typename BaseType::InputType>(input), callback);
+    }
+
+    void flush_impl(typename BaseType::CallbackFunction const & callback) override {
+        std::int64_t _numPending = 0;
+        while (_numPending < _max_future_offset) {
+            execute_helper(Traits<nonstd::optional<InputT>>::CreateNullValue(), callback);
+            ++_numPending;
+        }
+    }
+
+    void execute_helper(nonstd::optional<typename BaseType::InputType> const &input, typename BaseType::CallbackFunction const &callback);
+};
+
+/////////////////////////////////////////////////////////////////////////
+///  \typedef       LagLeadOperatorEstimator
+///  \brief         Estimator that creates `LagLeadOperatorTransformer`.
+///
+template <typename InputT>
+using LagLeadOperatorEstimator           = Components::InferenceOnlyEstimatorImpl<LagLeadOperatorTransformer<InputT>>;
+
+// ----------------------------------------------------------------------
+// ----------------------------------------------------------------------
+// ----------------------------------------------------------------------
+// |
+// |  Implementation
+// |
+// ----------------------------------------------------------------------
+// ----------------------------------------------------------------------
+// ----------------------------------------------------------------------
+template <typename T>
+LagLeadOperatorTransformer<T>::LagLeadOperatorTransformer(std::uint32_t horizon, std::vector<std::int64_t> offsets) :
+    _horizon(
+        std::move(
+            [&horizon](void) -> std::uint32_t & {
+                if (horizon == 0) {
+                    throw std::invalid_argument("Horizon cannot be 0!");
+                }
+                return horizon;
+            }()
+    )),
+    _offsets(
+        std::move(
+            [&offsets](void) -> std::vector<std::int64_t> & {
+                if (offsets.size() == 0) {
+                    throw std::invalid_argument("Offsets is empty!");
+                }
+                return offsets;
+            }()
+        )
+    ),
+    _max_future_offset(std::move(
+        *std::max_element(_offsets.cbegin(), _offsets.cend()) > 0 ?  *std::max_element(_offsets.cbegin(), _offsets.cend()) : 0
+    )),
+    _max_prior_offset(std::move(
+        *std::min_element(_offsets.cbegin(), _offsets.cend()) < 0 ?  *std::min_element(_offsets.cbegin(), _offsets.cend()) : 0
+    )),
+    // if there are more than one offset provided, we need find the difference between max and min value to determine buffer size
+    // and if there's only one offset, we use the absolute value of it as the size
+    _buffer(
+        _horizon + 
+        (_offsets.size() == 1 ? 
+        static_cast<size_t>(std::abs(_offsets[0])) : 
+        static_cast<size_t>(_max_future_offset-_max_prior_offset))
```

Comment:
This check doesn't work in all scenarios; for example, what happens if the caller incorrectly passes in `[-2, -2]` or `[0]`. This needs to be more resilient to handle all scenarios. Also, tests need to be added to verify these scenarios.

Reply:
Changed the implementation and added unit tests for these scenarios.